pieforproviders/pieforproviders
# Feature/54/form validation

# 💅🏼 What issue does this fix?

Part of a series of PRs for issue #54 

This handles form validation using the `react-hook-form` library, and in the process:
* changes the dropdown and radio inputs back to native elements
* gets rid of the obnoxious phone input library I had initially
* lots of param tweaks to clarify usage of the reusable form components

## 🍂 Before You Submit

  * I did not write tests, specifically because this is almost all reliant on basic html element/dom element behavior; I'm trying to avoid testing the libraries themselves rather than real functionality

## 🛷 Deployment Considerations

  * `react-hook-form`

## 🧵 Steps to set up locally

* `yarn install`

## 🧳 Steps to test

* visit `localhost:3000/signup`; you should see that the form renders correctly
* scroll to the bottom of the page and try to click the Submit button; you should not be able to because the button is disabled
* put your cursor in the full name box and tab out; you should see an error message telling you the full name is required (you can repeat this for all the rest of the validations required by the ticket)
 
## 🕯 Caveats, concerns
It's unclear to me if these inputs are controlled or uncontrolled at this point, I've been round and round on it so many times.  I'd _love_ to refactor so they're all controlled components but I'm not sure it's in scope or that I have the context to do that with hand-rolled form components and this `react-hook-form` validation library.

This Signup.js component is the worst; I think this happens a lot with components that handle forms and other "interactive" elements but I wish I knew how to refactor it into smaller chunks.  It's a lot to maintain, but the idea of building a component for every signup form input field sounds even worse.

This does NOT handle API calls; that will come in the next PR.
